sdks/python/apache_beam/io/gcp/bigquery_tools.py

```diff
@@ -998,6 +1004,13 @@ def encode(self, table_row):
     # This code will catch this error to emit an error that explains
     # to the programmer that they have used NAN/INF values.
     try:
+      # on python 3 base64 bytes are decoded to strings before being send to bq
+      if sys.version[0] == '3':
+        if type(table_row) == str:
```

Comment:
When do we go through this branch? Why do we only go through json.loads() here on Python 3, but not Python 2?

Reply:
This codepath will decode bytes when they are written to bigquery (it expects bytes to be base-64 encoded but allows them to have type `bytes` or `str`) This is to allow the same data that is read by bq to also be writen to bq.

in https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/gcp/bigquery_file_loads_test.py#L54 data is written to bigquery as a string so that is why we first do a `json.loads ()`before handling the bytes decoding (and this is only necessary in python 3).

Reply:
I still don't understand why/when we may receive a string here and not a dictionary (as per docstring) during `encode()` call. The test you referenced seems to call json.loads() before ingesting data into BQ: https://github.com/apache/beam/blob/23115ce75c296def30c304e3b386632c270ba224/sdks/python/apache_beam/io/gcp/bigquery_file_loads_test.py#L83. I created a PR on top of yours to see if the test pass without this change: https://github.com/apache/beam/pull/8709, waiting for the tests to finish.

Reply:
https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/gcp/bigquery_file_loads_test.py#L244. In this test no `json.loads()` is done. Should I edit the tests to pass dicts instead of strings?

Reply:
I think that the test needs to be fixed to pass dicts or BQ IO file loads method needs to use a different codec, not a codec that expects dicts. 
I suggest we change the test to pass dicts.  @pabloem, @chamikaramj - do you have a different opinion on this?